libs/auth/src/java_card.ts

```diff
@@ -0,0 +1,462 @@
+ */
+function isIncorrectPinResponseApdu(statusWord: [Byte, Byte]): boolean {
+  const [sw1, sw2] = statusWord;
+  // eslint-disable-next-line no-bitwise
+  return sw1 === 0x63 && sw2 >> 4 === 0x0c;
+}
+
+/**
+ * See isIncorrectPinResponseApdu().
+ */
+function numRemainingAttemptsFromIncorrectPinResponseApdu(
+  statusWord: [Byte, Byte]
+): number {
+  assert(isIncorrectPinResponseApdu(statusWord));
+  const [, sw2] = statusWord;
+  // Extract the last 4 bits of SW2
+  // eslint-disable-next-line no-bitwise
+  return sw2 & 0x0f;
+}
+
+/**
+ * An implementation of the card API that uses a Java Card running our fork of the OpenFIPS201
+ * applet (https://github.com/votingworks/openfips201) and X.509 certs. The implementation takes
+ * inspiration from the NIST PIV standard but diverges where PIV doesn't suit our needs.
+ */
+export class JavaCard implements Card {
+  private readonly cardReader: CardReader;
+  private cardStatus: CardStatus;
+  private readonly jurisdiction: string;
+  private readonly vxCertAuthorityCertPath: string;
+
+  constructor(input: {
+    jurisdiction: string;
+    vxCertAuthorityCertPath: string;
+  }) {
+    this.cardStatus = { status: 'no_card' };
+    this.jurisdiction = input.jurisdiction;
+    this.vxCertAuthorityCertPath = input.vxCertAuthorityCertPath;
+
+    this.cardReader = new CardReader({
+      onReaderStatusChange: async (readerStatus) => {
+        switch (readerStatus) {
+          case 'error': {
+            this.cardStatus = { status: 'error' };
+            return;
+          }
+          case 'no_card':
+          case 'no_card_reader': {
+            this.cardStatus = { status: 'no_card' };
+            return;
+          }
+          case 'ready': {
+            const user = await this.safeReadUser();
+            // Verify that the card is still in the card reader before updating the status since
+            // the async user-reading operation can take time and the card may have been removed
+            // since the operation started
+            if (this.cardReader.getReaderStatus() === 'ready') {
+              this.cardStatus = { status: 'ready', user };
+            }
+            return;
+          }
+          default: {
+            throwIllegalValue(readerStatus);
+          }
+        }
+      },
+    });
+  }
+
+  async getCardStatus(): Promise<CardStatus> {
+    return Promise.resolve(this.cardStatus);
+  }
+
+  async checkPin(pin: string): Promise<CheckPinResponse> {
+    const cardVxAdminCert = await this.retrieveCert(
+      CARD_VX_ADMIN_CERT.OBJECT_ID
+    );
+    try {
+      // Verify that the card has a private key that corresponds to the public key in the card
+      // VxAdmin cert
+      await this.verifyCardPrivateKey(
+        CARD_VX_ADMIN_CERT.PRIVATE_KEY_SLOT,
+        cardVxAdminCert,
+        pin
+      );
+    } catch (error) {
+      if (
+        error instanceof ResponseApduError &&
+        isIncorrectPinResponseApdu(error.statusWord())
+      ) {
```

Comment:
At a glance, one might see this use case as a reason to have the underlying `CardReader.transmit` method return a `Result` object. I initially tried that but found it clunky to have to add the following check in so many places and through so many layers of helpers when in most cases I just wanted errors to bubble up:

```ts
if (!response.isOk()) {
  return response;
}
```

That's why I settled on the `ResponseApduError` concept.

I'm increasingly seeing the `Result` object's primary use case as the interface between the backend and the frontend. I find errors to be more convenient within the backend.

Reply:
That makes a lot of sense! I'm going to keep my logic as I have it for now but will keep an open mind for other cases!